src/evaluators/ForOfStatement.js

```diff
@@ -163,13 +163,8 @@ export function ForInOfHeadEvaluation(
     let obj = ToObjectPartial(realm, exprValue);
 
     // c. Return ? EnumerateObjectProperties(obj).
-    if (obj instanceof AbstractObjectValue) {
-      if (obj.isSimple()) {
-        return obj;
-      } else {
-        obj.throwIfNotConcrete();
-      }
-      invariant(false);
+    if (obj instanceof AbstractValue) {
```

Comment:
Let the caller deal with giving an error, since the caller has more context to explain why.